internal/controllers/cluster/cluster_controller_phases.go

```diff
@@ -50,10 +50,16 @@ func (r *Reconciler) reconcilePhase(_ context.Context, cluster *clusterv1.Cluste
 
 	if cluster.Spec.InfrastructureRef != nil {
 		cluster.Status.SetTypedPhase(clusterv1.ClusterPhaseProvisioning)
```

Comment:
Can we set the phase to ClusterPhaseProvisioning if one of the refs is set, and then later set it to provisioned if all control plane & infrastructure objects that are referenced are ready + the endpoint is valid?

Reply:
Wondering if we also have to set the infrastructureReady field to false

> - All the controllers working with Cluster objects must take into account that the `infrastructureRef` field could be omitted; most notably:
>  - The Cluster controller must use skip reconciling this external reference when the `infrastructureRef` is missing; also, the `status.InfrastructureReady` field must be automatically set to true in this case.

Reply:
I guess with that change to reconcile infrastructure, we could use the following here

```go
	if cluster.Spec.InfrastructureRef != nil || cluster.Spec.ControlPlaneRef != nil {
		cluster.Status.SetTypedPhase(clusterv1.ClusterPhaseProvisioning)
	}

	if cluster.Status.InfrastructureReady && cluster.Status.ControlPlaneReady && cluster.Spec.ControlPlaneEndpoint.IsValid() {
		cluster.Status.SetTypedPhase(clusterv1.ClusterPhaseProvisioned)
	}
```